week4/trie/thalesmg/Main.hs

```diff
@@ -0,0 +1,51 @@
+import qualified Data.Map.Lazy as M
+
+data Trie a = Empty | Node (M.Map a (Trie a)) deriving (Show)
+
+mkTrie :: [a] -> Trie a
+mkTrie [] = Empty
+mkTrie (c:cs) = Node (M.singleton c (mkTrie cs))
+
+addToTrie :: (Ord a) => Trie a -> [a] -> Trie a
+addToTrie t [] = t
+addToTrie Empty xs = mkTrie xs
+addToTrie (Node mp) (x:xs) = 
+  case M.lookup x mp of
+    Nothing -> Node (M.insert x (mkTrie xs) mp)
+    Just t' -> Node (M.insert x (addToTrie t' xs) mp)
+    
+countNodes :: Trie a -> Int
+countNodes t =
```

Comment:
I think using sum and map you get a slightly more readable version:

countNodes2 :: Trie a -> Int
countNodes2 Empty = 0
countNodes2 (Node mp) = M.size mp + sum (M.map countNodes2 mp)